### PR TITLE
ECDC-3965_da00_serializer

### PR DESCRIPTION
**Pull request by tbukovics**
### Issue reference / description

Subtast of the CBM to create a new serializer. This PR is for the follow up of the new serilazer concept as well as the new serializer.

## Checklist for submitter

- [x] Check for conflict with integration test
- [x] Unit tests pass

---

### Nominate for Group Code Review

- [x] Nominate for code review

### CODE REVIEW

**Comment by tbukovics** [comment]
Hi @mortenjc 

You can have an early look how the new serializer looks like. I disused this with Greg and I mostly based everything on his code. I reworked a bit on the concept how we organize Serializers in our code. Currently the da00 is the first following this concept but I planning to rework the other two as well if this goes well.

File: src/common/kafka/serializer/AbstractSerializer.h [hunk 1]
```diff
@@ -14,15 +14,29 @@
 //===----------------------------------------------------------------------===//
 
 #include "common/kafka/Producer.h"
```
Comment (mortenjc) [inline src/common/kafka/serializer/AbstractSerializer.h:16]: <common/kafka/Producer.h> ;-)

File: src/common/kafka/serializer/BinAggregation.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright (C) 2024 European Spallation Source, ERIC. see LICENSE file
```
Comment (mortenjc) [inline src/common/kafka/serializer/BinAggregation.h:1]: This seem to be a generic math abstraction. Maybe it should belong to a 
math library for example <common/math>  ?
And then not using the serializer namespace but essmath namespace or something?

File: src/common/kafka/serializer/FlatbufferTypes.h [hunk 3]
```diff
@@ -267,4 +265,4 @@ class DataArray {
   }
 };
 
-}; // namespace da00_serializer
\ No newline at end of file
+}; // namespace da00_faltbuffers
```
Comment (mortenjc) [inline src/common/kafka/serializer/FlatbufferTypes.h:268]: flatbuffers

File: src/common/kafka/serializer/FlatbufferTypes.h [hunk 4]
```diff
@@ -16,15 +16,13 @@
...
 #include <chrono>
-#include <map>
+#include <da00_dataarray_generated.h>
 #include <numeric>
 #include <optional>
 #include <sstream>
-#include <string>
-#include <da00_dataarray_generated.h>
 
 namespace da00_faltbuffers {
```
Comment (mortenjc) [inline src/common/kafka/serializer/FlatbufferTypes.h:24]: flatbuffers

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+/// \file
+///
+/// \brief Implementation of Frame1DHistogramSerializer
+///
+/// See https://github.com/ess-dmsc/streaming-data-types
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "AbstractSerializer.h"
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:13]: #include <>

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+};
+template <> struct data_type_trait<float> {
+  static constexpr da00_dtype type = da00_dtype::float32;
+};
+template <> struct data_type_trait<double> {
+  static constexpr da00_dtype type = da00_dtype::float64;
+};
+
+/// \brief A class to handle sending 1-D data collected over a frame to Kafka
+/// using da00 flatbuffers schema
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:59]: Maybe describe the function of R and T?

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+template <class T, class R = T>
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:60]: General note: This file deviates from the ECDC coding standard regarding variable naming.
The current code standard requires variable names to be CamelCase. So 'topic', 'minValue' and '_timeUnit' are 
in violation.

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+class Frame1DHistogramBuilder : public AbstractSerializer {
+  using data_t = std::vector<std::vector<T>>;
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:62]: Not sure I understand why we use a vector of vectors here?

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+  /// \brief Construct a Frame1DHistogramSender
+  Frame1DHistogramBuilder(
+      std::string topic, //!< Kafka stream topic destination
+      const time_t
+          period, //!< length of time of one frame in the specified units
+      const time_t binCount, //!< number of time bins used
+      std::string name,      //!< name of the binned data
+      std::string unit,      //!< unit of the binned data
+      std::string timeUnit =
+          "millisecond", //!< unit of time used for period, and the binned axis
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:85]: Great to see some comments. But I would prefer these to be above the constructor for readability

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+      const AggreggeFunc<T> &aggFunc = SUM_AGG_FUNC<T>)
+      : AbstractSerializer(callback), _topic(std::move(topic)), _period(period),
+        _binCount(binCount), _name(std::move(name)), _unit(std::move(unit)),
+        _timeUnit(std::move(timeUnit)), aggBinData(aggFunc) {
+
+    initAxis();
+    _data = data_t(_xAxis.size());
+  }
+
+  inline void addData(const time_t rToA, const T data) {
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:96]: I believe that rToA is misleading. Isn't it really TOF? and what does 'r' stand for?

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h (hunk 8 above)
Comment (mortenjc) [reply src/common/kafka/serializer/Frame1DHistogramSerializer.h:62]: I understand that you (dynamically) allocate vectors for each bin. But why do this when we know that all we need to store is the accumulated value?

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    // requires that initXaxis create an ascending sorted vector
+    R minValue = _xAxis.front();                          // minimum value
+    R maxValue = _xAxis.back();                           // maximum value
+    R step = (maxValue - minValue) / (_xAxis.size() - 1); // step size
+
+    int binIndex = static_cast<int>((rToA - minValue) / step);
+
+    // Reserve space for the data for optimize on performance
+    if (_data[binIndex].empty())
+      _data[binIndex].reserve(50);
```
Comment (mortenjc) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:108]: 50 seem a bit arbitrary to me.

**Review by mortenjc — CHANGES_REQUESTED** [review]
Good work and good ideas.

There is a bit of #include "" and variable naming that should be made to fit the
code standard and the current practices. I think some of the code is needlessly complicated. For example having these template functions to return a sum or an average. Also I don't like the vector of vectors due to the dynamic allocation and freeing of data that it implies. For this use case  all we really want is to accumulate the counts in every bin which can be supported by a single preallocated vector.

But we can discuss this tomorrow (Tuesday).

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h (hunk 9 above)
Comment (tbukovics) [reply src/common/kafka/serializer/Frame1DHistogramSerializer.h:85]: OK, I will add comments there as well

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+                  .data(_xAxis));
+
+    auto y = (Variable(_name, {"t"}, {static_cast<time_t>(_binCount)})
+                  .unit(_unit)
+                  .source("histogram_sender")
+                  .label("Frame 1D Histogram intensity")
+                  .data(aggregatedBins));
+
+    const auto dataarray = DataArray("histogram_sender", {x, y});
+
```
Comment (tbukovics) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:181]: THe source here should be part of the constructor

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+  /// \brief Serialize the data to a flatbuffer.
+  /// \details The data is serialized to a flatbuffer and stored in the buffer
+  /// of the abstract serializer
+  void serialize() {
+    if (static_cast<time_t>(_data.size()) != _binCount) {
+      std::stringstream ss;
+      ss << "Expected data to serialize to have " << _binCount
+         << " elements but provided " << _data.size();
+      throw std::runtime_error(ss.str());
+    }
```
Comment (tbukovics) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:160]: Use the fmt to create the output message everywhere

File: src/common/kafka/serializer/Frame1DHistogramSerializer.h [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+        _binCount(binCount), _name(std::move(name)), _unit(std::move(unit)),
+        _timeUnit(std::move(timeUnit)), aggregateFunction(aggFunc) {
+
+    initAxis();
+    _data = data_t(_xAxis.size());
+  }
+
+  inline void addData(const R rToA, const T data) {
+    static_assert(data_type_trait<R>::type != da00_dtype::none,
+                  "Data type R not supported for serialization!");
```
Comment (tbukovics) [inline src/common/kafka/serializer/Frame1DHistogramSerializer.h:120]: Pease add here an config parameter which setup the binning strategy for data which out of the time scope, like add last bin or drop

**Review by mortenjc — APPROVED** [review]

**Pull request merged**